### PR TITLE
tree2: Remove support for global fields

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Support for global fields has been removed.

There is still a root field schema, but it is now directly on SchemaData instead of part of the global field map.

Additionally, some related changes have been made, mainly:
1. forbiddenFieldKindIdentifier has been added to core, removing the need to inject emptyField from schema policy.
2. Unrecognized tree schema identifiers are no longer implicitly filled with a default tree schema (typically never schema), and thus no longer need this default injected via schema policy.
3. Due to the above two items, schema policy is no longer needed except for listing field kinds, and many places just use SchemaData instead now.

Simplifications which are easy to separate and leave for later have not been done yet (for example merging FieldKey and LocalFieldKey) to keep the size of this PR smaller.

## Breaking Changes

Anything using global fields or SchemaData is broken, and may need to be updated similar to the tests in this change.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

**Review by jenn-le — APPROVED** [review]

File: experimental/dds/tree2/src/feature-libraries/chunked-forest/chunkedForest.ts [hunk 1]
```diff
@@ -55,25 +55,23 @@ class ChunkedForest extends SimpleDependee implements IEditableForest {
 
 	public constructor(
 		public roots: BasicChunk,
-		public readonly schema: StoredSchemaRepository<FullSchemaPolicy>,
+		public readonly schema: StoredSchemaRepository,
+		public readonly policy: FullSchemaPolicy,
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/chunked-forest/chunkedForest.ts:65]: no real changes here. I just removed policy from StoredSchemaRepository so its now stored separate here. It just holds the map of field kinds. I've added some more docs for these.

File: experimental/dds/tree2/src/feature-libraries/chunked-forest/chunkedForest.ts [hunk 2]
```diff
@@ -53,6 +53,12 @@ class ChunkedForest extends SimpleDependee implements IEditableForest {
 
 	private readonly events = createEmitter<ForestEvents>();
 
+	/**
+	 * @param roots - dummy node above the root under which detached fields are stored. All content of the forest is reachable from this.
+	 * @param schema - schema which all content is this forest is assumed to comply with.
+	 * @param policy - provides information needed to interpret the schema, mainly multiplicities for each field kind whish are used for chunking policy.
+	 * @param anchors - anchorSet use to track location in this forest across changes. Callers of applyDelta must ensure this is updated this accordingly.
```
Comment (jenn-le) [inline experimental/dds/tree2/src/feature-libraries/chunked-forest/chunkedForest.ts:60]:
```suggestion
	 * @param schema - schema which all content in this forest is assumed to comply with.
	 * @param policy - provides information needed to interpret the schema, mainly multiplicities for each field kind which are used for chunking policy.
	 * @param anchors - anchorSet used to track location in this forest across changes. Callers of applyDelta must ensure this is updated accordingly.
```

File: experimental/dds/tree2/src/feature-libraries/typed-schema/buildViewSchemaCollection.ts [hunk 3]
```diff
@@ -82,6 +82,12 @@ export function buildViewSchemaCollection(
 	}
 
 	return {
+		// To implement the SchemaData interface (which SchemaCollection extends), a rootFieldSchema.
+		// In the case where this schema collection is a library and not full document schema, no rootFieldSchema:
+		// in this case comparability with SchemaData can be enabled by adding a rootFieldSchema.
+		// For this case, a forbidden field is used,
```
Comment (jenn-le) [inline experimental/dds/tree2/src/feature-libraries/typed-schema/buildViewSchemaCollection.ts:88]: This was a bit difficult to read.

**Comment by alexvy86** [comment]
I can reproduce the error in a codespaces instance with this branch, and not locally. The error: `"Uncaught ReferenceError: Cannot access 'value' before initialization", source: http://localhost:8229/app.bundle.js (5981)`

<details>
<summary>Output with this branch</summary>

```
[16][15:00:48] npm run test:jest:verbose

> @fluid-internal/devtools-browser-extension@2.0.0-internal.5.4.0 test:jest:verbose
> cross-env FLUID_TEST_VERBOSE=1 jest

Determining test suites to run...Fontconfig warning: "/usr/share/fontconfig/conf.avail/05-reset-dirs-sample.conf", line 6: unknown element "reset-dirs"

DevTools listening on ws://127.0.0.1:39093/devtools/browser/24e6fb34-1b64-42a5-8459-3dead5edb416
[0727/150049.467540:ERROR:command_buffer_proxy_impl.cc(124)] ContextResult::kTransientFailure: Failed to send GpuChannelMsg_CreateCommandBuffer.
<i> [webpack-dev-server] Project is running at:
<i> [webpack-dev-server] Loopback: http://localhost:8229/
<i> [webpack-dev-server] On Your Network (IPv4): http://172.16.5.4:8229/
<i> [webpack-dev-server] Content not from webpack is served from 'e2e-tests/app' directory
<i> [webpack-dev-middleware] wait until bundle finished: /
[0727/150058.080638:INFO:CONSOLE(113019)] "[HMR] Waiting for update signal from WDS...", source: http://localhost:8229/app.bundle.js (113019)
[0727/150058.111736:INFO:CONSOLE(100153)] "%cDownload the React DevTools for a better development experience: https://reactjs.org/link/react-devtools font-weight:bold", source: http://localhost:8229/app.bundle.js (100153)
[0727/150058.163389:INFO:CONSOLE(5981)] "Uncaught ReferenceError: Cannot access 'value' before initialization", source: http://localhost:8229/app.bundle.js (5981)
[0727/150058.165297:INFO:CONSOLE(111847)] "[webpack-dev-server] Hot Module Replacement enabled.", source: http://localhost:8229/app.bundle.js (111847)
[0727/150058.165452:INFO:CONSOLE(111847)] "[webpack-dev-server] Live Reloading enabled.", source: http://localhost:8229/app.bundle.js (111847)
[0727/150058.203635:INFO:CONSOLE(113019)] "[HMR] Waiting for update signal from WDS...", source: http://localhost:8229/app.bundle.js (113019)
[0727/150058.206045:INFO:CONSOLE(100153)] "%cDownload the React DevTools for a better development experience: https://reactjs.org/link/react-devtools font-weight:bold", source: http://localhost:8229/app.bundle.js (100153)
[0727/150058.218184:INFO:CONSOLE(5981)] "Uncaught ReferenceError: Cannot access 'value' before initialization", source: http://localhost:8229/app.bundle.js (5981)

 RUNS  e2e-tests/chromebrowserdevtool.test.ts
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error: Navigation failed because browser has disconnected!
    at CDPSession.<anonymous> (/workspaces/FluidFramework/node_modules/.pnpm/puppeteer@1.20.0/node_modules/puppeteer/lib/LifecycleWatcher.js:46:107)
    at CDPSession.emit (node:events:513:28)
    at CDPSession._onClosed (/workspaces/FluidFramework/node_modules/.pnpm/puppeteer@1.20.0/node_modules/puppeteer/lib/Connection.js:215:10)
    at Connection._onMessage (/workspaces/FluidFramework/node_modules/.pnpm/puppeteer@1.20.0/node_modules/puppeteer/lib/Connection.js:105:17)
    at WebSocket.<anonymous> (/workspaces/FluidFramework/node_modules/.pnpm/puppeteer@1.20.0/node_modules/puppeteer/lib/WebSocketTransport.js:44:24)
    at WebSocket.onMessage (/workspaces/FluidFramework/node_modules/.pnpm/ws@6.2.2/node_modules/ws/lib/event-target.js:120:16)
    at WebSocket.emit (node:events:513:28)
    at Receiver.receiverOnMessage (/workspaces/FluidFramework/node_modules/.pnpm/ws@6.2.2/node_modules/ws/lib/websocket.js:789:20)
    at Receiver.emit (node:events:513:28)
    at Receiver.dataMessage (/workspaces/FluidFramework/node_modules/.pnpm/ws@6.2.2/node_modules/ws/lib/receiver.js:422:14)
  -- ASYNC --
    at Frame.<anonymous> (/workspaces/FluidFramework/node_modules/.pnpm/puppeteer@1.20.0/node_modules/puppeteer/lib/helper.js:111:15)
    at Page.goto (/workspaces/FluidFramework/node_modules/.pnpm/puppeteer@1.20.0/node_modules/puppeteer/lib/Page.js:674:49)
    at Page.<anonymous> (/workspaces/FluidFramework/node_modules/.pnpm/puppeteer@1.20.0/node_modules/puppeteer/lib/helper.js:112:23)
    at Object.<anonymous> (/workspaces/FluidFramework/packages/tools/devtools/devtools-browser-extension/e2e-tests/chromebrowserdevtool.test.ts:42:14)
    at Object.asyncJestLifecycle (/workspaces/FluidFramework/node_modules/.pnpm/jest-jasmine2@26.6.3/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:58:37)
    at /workspaces/FluidFramework/node_modules/.pnpm/jest-jasmine2@26.6.3/node_modules/jest-jasmine2/build/queueRunner.js:45:12
    at new Promise (<anonymous>)
    at mapper (/workspaces/FluidFramework/node_modules/.pnpm/jest-jasmine2@26.6.3/node_modules/jest-jasmine2/build/queueRunner.js:28:19)
    at /workspaces/FluidFramework/node_modules/.pnpm/jest-jasmine2@26.6.3/node_modules/jest-jasmine2/build/queueRunner.js:75:41
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```
</details>

<details>
<summary>Output with main</summary>

```
[22][10:11:38] npm run test:jest:verbose

> @fluid-internal/devtools-browser-extension@2.0.0-internal.5.4.0 test:jest:verbose
> cross-env FLUID_TEST_VERBOSE=1 jest

Determining test suites to run...
DevTools listening on ws://127.0.0.1:38463/devtools/browser/eeedb696-65e7-47bf-92b1-af77c0544ea6
[0727/101139.561013:ERROR:command_buffer_proxy_impl.cc(124)] ContextResult::kTransientFailure: Failed to send GpuChannelMsg_CreateCommandBuffer.
<i> [webpack-dev-server] Project is running at:
<i> [webpack-dev-server] Loopback: http://localhost:8229/
<i> [webpack-dev-server] On Your Network (IPv4): http://172.29.143.40:8229/
<i> [webpack-dev-server] Content not from webpack is served from 'e2e-tests/app' directory
<i> [webpack-dev-middleware] wait until bundle finished: /
[0727/101148.869685:INFO:CONSOLE(113253)] "[HMR] Waiting for update signal from WDS...", source: http://localhost:8229/app.bundle.js (113253)
[0727/101148.910151:INFO:CONSOLE(100387)] "%cDownload the React DevTools for a better development experience: https://reactjs.org/link/react-devtools font-weight:bold", source: http://localhost:8229/app.bundle.js (100387)
[0727/101149.054298:INFO:CONSOLE(160446)] "Fluid Devtools: Posting messages to the window: [object Object]", source: http://localhost:8229/app.bundle.js (160446)
[0727/101149.105013:INFO:CONSOLE(160446)] "Fluid Devtools: Posting messages to the window: [object Object]", source: http://localhost:8229/app.bundle.js (160446)
[0727/101149.105209:INFO:CONSOLE(160446)] "Fluid Devtools: Posting messages to the window: [object Object]", source: http://localhost:8229/app.bundle.js (160446)
[0727/101149.128773:INFO:CONSOLE(108416)] "App rendered!", source: http://localhost:8229/app.bundle.js (108416)
[0727/101149.135738:INFO:CONSOLE(112081)] "[webpack-dev-server] Hot Module Replacement enabled.", source: http://localhost:8229/app.bundle.js (112081)
[0727/101149.135897:INFO:CONSOLE(112081)] "[webpack-dev-server] Live Reloading enabled.", source: http://localhost:8229/app.bundle.js (112081)
[0727/101149.163606:INFO:CONSOLE(113253)] "[HMR] Waiting for update signal from WDS...", source: http://localhost:8229/app.bundle.js (113253)
[0727/101149.167194:INFO:CONSOLE(100387)] "%cDownload the React DevTools for a better development experience: https://reactjs.org/link/react-devtools font-weight:bold", source: http://localhost:8229/app.bundle.js (100387)
[0727/101149.215355:INFO:CONSOLE(160446)] "Fluid Devtools: Posting messages to the window: [object Object]", source: http://localhost:8229/app.bundle.js (160446)
[0727/101149.343206:INFO:CONSOLE(108416)] "App rendered!", source: http://localhost:8229/app.bundle.js (108416)
[0727/101149.345305:INFO:CONSOLE(160446)] "Fluid Devtools: Posting messages to the window: [object Object]", source: http://localhost:8229/app.bundle.js (160446)
[0727/101149.345936:INFO:CONSOLE(112081)] "[webpack-dev-server] Hot Module Replacement enabled.", source: http://localhost:8229/app.bundle.js (112081)
[0727/101149.355676:INFO:CONSOLE(112081)] "[webpack-dev-server] Live Reloading enabled.", source: http://localhost:8229/app.bundle.js (112081)
 PASS  e2e-tests/chromebrowserdevtool.test.ts (7.982 s)
  End to end tests
    ✓ Smoke: verify test app can be launched (271 ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        8.05 s
Ran all test suites.
<i> [webpack-dev-server] Gracefully shutting down. To force exit, press ^C again. Please wait...
```
</details>

Running `npm run start:test:client` for the devtools-browser-extension on this branch and going to http://localhost:8080 I could get a better stack trace in the console:

```
bootstrap:27  Uncaught ReferenceError: Cannot access 'value' before initialization
    at Module.value (defaultFieldChangeFormat.js:67:1)
    at ../../../../experimental/dds/tree2/lib/feature-libraries/default-field-kinds/defaultSchema.js (defaultSchema.js:10:39)
    at __webpack_require__ (bootstrap:24:1)
    at fn (hot module replacement:62:1)
    at ../../../../experimental/dds/tree2/lib/feature-libraries/modular-schema/comparison.js (mapTreeCursor.js:37:1)
    at __webpack_require__ (bootstrap:24:1)
    at fn (hot module replacement:62:1)
    at ../../../../experimental/dds/tree2/lib/feature-libraries/modular-schema/fieldKind.js (fieldChangeHandler.js:35:1)
    at __webpack_require__ (bootstrap:24:1)
    at fn (hot module replacement:62:1)
```

As far as I'm aware, the test app for these jest tests does not use SharedTree, it's only a transitive dependency... do we expect to have to make any changes there?

@jikim-msft @Josmithr @chentong7 FYI

**Comment by Josmithr** [comment]
> I can reproduce the error in a codespaces instance with this branch, and not locally. The error: `"Uncaught ReferenceError: Cannot access 'value' before initialization", source: http://localhost:8229/app.bundle.js (5981)`
> 
> Output with this branch
> Output with main
> Running `npm run start:test:client` for the devtools-browser-extension on this branch and going to http://localhost:8080 I could get a better stack trace in the console:
> 
> ```
> bootstrap:27  Uncaught ReferenceError: Cannot access 'value' before initialization
>     at Module.value (defaultFieldChangeFormat.js:67:1)
>     at ../../../../experimental/dds/tree2/lib/feature-libraries/default-field-kinds/defaultSchema.js (defaultSchema.js:10:39)
>     at __webpack_require__ (bootstrap:24:1)
>     at fn (hot module replacement:62:1)
>     at ../../../../experimental/dds/tree2/lib/feature-libraries/modular-schema/comparison.js (mapTreeCursor.js:37:1)
>     at __webpack_require__ (bootstrap:24:1)
>     at fn (hot module replacement:62:1)
>     at ../../../../experimental/dds/tree2/lib/feature-libraries/modular-schema/fieldKind.js (fieldChangeHandler.js:35:1)
>     at __webpack_require__ (bootstrap:24:1)
>     at fn (hot module replacement:62:1)
> ```
> 
> As far as I'm aware, the test app for these jest tests does not use SharedTree, it's only a transitive dependency... do we expect to have to make any changes there?
> 
> @jikim-msft @Josmithr @chentong7 FYI

The view / extension don't do anything with any of the DDSs, but they use `devtools-core`, which includes the default visualizers for our various DDSs. Seems like webpack isn't giving us the tree shaking benefits I would hope for. Probably worth an explicit investigation. I would bet the extension's bundle is way bigger than it needs to be too.

**Comment by jenn-le** [comment]
[celebrate]     Jenn Le reacted to your message:
________________________________

**Pull request merged**